src/server/server_state.h

```diff
@@ -267,6 +267,14 @@ class ServerState {  // public struct - to allow initialization.
   // Exec descriptor frequency count for this thread.
   absl::flat_hash_map<std::string, unsigned> exec_freq_count;
 
+  struct CurrentSaveStats {
+    // snapshot related
+    size_t keys_processed = 0;
+    size_t keys_total = 0;
+  };
+
+  CurrentSaveStats current_save_stats;
```

Comment:
why dont you put this stats in SliceSnapshot class and add a funciton in SaveStagesController that aggregates the stats as we do in GetSaveBuffersSize

Reply:
if you have a new function GetSaveProgress you will not need the save_mode_ as well. you can always set the counters and reply them in info persistance only if save_controller_ exists

Reply:
The thing is that we need `save_mode_` because otherwise a new `SliceSnapshot` (that is not part of the bgsave/save operation) might overwrite the thread local. We exercise almost the same path in replication, so it could be the case that replication and a bgsave operation happen at the same time. Both of the two command paths will happily increment the thread locals messing with the correctness.

Second is that with what I have here, we save a hop. We don't dispatch a cb to fetch and aggregate the thread locals. We are in non critical path so this is less important but still

Reply:
I am suggesting to add the stats to SliceSnapshot class, not to use the thread local

Reply:
I did the change and to be honest I am not very happy with it:

1. It's more code and introduces a way more convoluted code path
2. I needed 3 different proxy functions to reach the `SliceSnapshot` :cry: 
3. code repetition because we need to handle explicitly different formats/requirements

Reply:
but it's fine for now, but at some point I would consider a refactor. It's too much work to plug in stat tracking. That 3 proxy call in `GetSCurrentSnapshotProgress` killed it for me

Reply:
https://github.com/dragonflydb/dragonfly/issues/2644

Reply:
I agree there are lots of proxy functions in this flow and it looks messy
we can check later if we can do some refactoring here to make it simpler